lib/sles4sap/database_hana.pm

```diff
@@ -0,0 +1,198 @@
+# SUSE's openQA tests
+#
+# Copyright 2017-2024 SUSE LLC
+# SPDX-License-Identifier: FSFAP
+# Maintainer: QE-SAP <qe-sap@suse.de>
+
+package sles4sap::database_hana;
+use strict;
+use warnings;
+use testapi;
+use Carp qw(croak);
+use Exporter qw(import);
+use saputils qw(check_crm_output get_primary_node get_failover_node calculate_hana_topology);
+use sles4sap::sapcontrol;
+
+our @EXPORT = qw(
+  hdb_stop
+  wait_for_failed_resources
+  wait_for_takeover
+  register_replica
+  get_node_roles
+  find_hana_resource_name
+);
+
+
+=head1 SYNOPSIS
+
+Package contains functions for interacting with hana database and related actions.
+
+=cut
+
+=head2 hdb_stop
+
+    hdb_stop(instance_id=>'00', [switch_user=>'sidadm', command=>'kill']);
+
+Stop hana database using C<HDB stop> command. Function expects to be executed as sidadm, however you can use B<switch_user>
+to execute command using C<sudo su -> as a different user. The user needs to have correct permissions for performing
+requested action.
+Function waits till all DB processes are stopped.
+
+=over
+
+=item * B<instance_id>: Database instance ID. Mandatory.
+
+=item * B<switch_user>: Execute command as specified user with help of C<sudo su ->. Default: undef
+
+=item * B<command>: HDB command to trigger. Default: stop
+
+=back
+
+=cut
+
+sub hdb_stop {
+    my (%args) = @_;
+    my $stop_timeout = 600;
+    $args{command} //= 'stop';
+    croak("Command '$args{command}' is not supported.") unless grep(/$args{command}/, ('kill', 'stop'));
+
+    $args{command} = 'kill -x' if ($args{command} eq 'kill');
+    my $sudo_su = $args{switch_user} ? "sudo su - $args{switch_user} -c" : '';
+    my $cmd = join(' ', $sudo_su, '"', 'HDB', $args{command}, '"');
+    record_info('HDB stop', "Executing '$cmd' on " . script_output('hostname'));
+    assert_script_run($cmd, timeout => $stop_timeout);
+
+    # Wait Hana processes to stop
+    sapcontrol_process_check(instance_id => $args{instance_id}, expected_state => 'stopped', wait_for_state => 'yes', timeout => $stop_timeout);
```

Comment:
I'd suggest to keep this one out from here, and in the caller. You never know how this function will be used in the future.

Reply:
I don't think this is a reason for removing such an important step from this function. If database has to be stopped function must ensure processes are down not leave it somewhere in a limbo. Leaving this to the caller is just asking for race conditions. 

You cannot tiptoe around the code because somebody might do something with it in the future.
If somebody does changes to the code it is their responsibility and responsibility of reviewers not to break anything with their PR. We do have many processes to block such actions. 
`sapcontrol_process_check` exists for more than a year without any changes beside move to another place.